pippy/count_flops.py

```diff
@@ -0,0 +1,160 @@
+
+torch.fx.Tracer.proxy_buffer_attributes = True
+
+PIPELINE = [
+    "torchscript-module-to-torch-backend-pipeline",
+    "torch-backend-to-linalg-on-tensors-backend-pipeline",
+    "func.func(linalg-bufferize)",
+    "func.func(convert-linalg-to-affine-loops)",
+    # The workhorse of this script - unroll "affine loops" with trip count less than 1000000 (i.e. all of them)
+    # completely.
+    "func.func(affine-loop-unroll{ unroll-full unroll-full-threshold=1000000 })",
+]
+
+
+def script_module_with_annotations(test_module, annotations):
+    class_annotator = ClassAnnotator()
+    recursivescriptmodule = torch.jit.script(test_module)
+    frozen_recursivescriptmodule = recursivescriptmodule
+    class_annotator.exportNone(frozen_recursivescriptmodule._c._type())
+    class_annotator.exportPath(frozen_recursivescriptmodule._c._type(), ["forward"])
+    class_annotator.annotateArgs(
+        frozen_recursivescriptmodule._c._type(), ["forward"], annotations
+    )
+    return frozen_recursivescriptmodule, class_annotator
+
+
+def compile_ts_module_to_mlir_module(mod, in_shapes):
+    recursivescriptmodule, class_annotator = script_module_with_annotations(
+        mod, [None] + [(in_shape, torch.float32, True) for in_shape in in_shapes]
+    )
+
+    mb = ModuleBuilder()
+    mb.import_module(recursivescriptmodule._c, class_annotator)
+    run_pipeline_with_repro_report(mb.module, ",".join(PIPELINE), "")
+
+    return mb.module
+
+
+PARTITION_COUNTER = 0
+
+
+def mod_partition(_node):
+    global PARTITION_COUNTER
+    partition = PARTITION_COUNTER
+    PARTITION_COUNTER += 1
+    return partition
+
+
+def find_op_in_submodule(submodule):
+    for node in submodule.graph.nodes:
+        if node.op in {"call_function", "call_module"}:
+            return node.name
+    return None
+
+
+def compile_model_op_by_op(model, example_input):
+    """
+    High-level this 
+    
+    1. Traces an nn.Module to get a fx graph
+    2. Splits the fx graph into submodules, one submodule per op
+    3. Runs shape propagation on the parent module so that each submodule has shape refined inputs; 
+       this shape refinement is necessary for lowering through torch-mlir
+    4. TS scripts each submodule (using the refined shapes to produce torch-mlir "annotations")
+    5. Compiles each TS module using torch-mlir/mlir **with full loop unrolling**
+    6. Map the mlir module back to the relevant identifiers (submodule name and aten op name)
+    
+    Returns a list of 3 tuples of (submodule name, aten op name, mlir module);  the mlir module
+    can then be passed to count_flop_latency_in_mlir_module to get a latency number, in terms of floating point ops.
+    """
+    tracer = torch.fx.Tracer()
+    graph = tracer.trace(model)
+    traced = torch.fx.GraphModule(model, graph)
+    module_with_submodules = split_module(traced, model, mod_partition)
+    global PARTITION_COUNTER
```

Comment:
Does this really need to be global? can it just be a closed-over function-local var? Otherwise this will have side effects when invoked multiple times

Reply:
good call